tidy3d/components/source.py

```diff
@@ -291,7 +301,72 @@ def amp_time(self, time: float) -> complex:
         return const * offset * oscillation * amp
 
 
-SourceTimeType = Union[GaussianPulse, ContinuousWave]
+class CustomSourceTime(Pulse):
+    """Custom source time dependence.
+
+    Note
+    ----
+    To interpolate to simulation time steps, the value provided at the nearest time is used.
+
+    Example
+    -------
+    >>> sourcetime = CustomSourceTime.from_numpy(values=np.linspace(0, 1, 100), dt=.1)
+    """
+
+    sourcetime_dataset: Optional[TimeDataset] = pydantic.Field(
```

Comment:
the `Source` has `source_time` as a field, so I'd vote we change this to `source_time_dataset` if it doesn't break any PEP8 convention for variable naming.